File: apps/api/app/api/model_control.py
from typing import Any

from fastapi import APIRouter

from app.model_state import get_model_state

router = APIRouter()


@router.get("/api/v1/model/state")
def active_model_state() -> dict[str, Any]:
    state = get_model_state()
    return {
        "banner": state.banner,
        "model": state.model.model_dump(),
        "rules": {
            "recent_total_leakage_confirmation": state.rules.recent_total_leakage_confirmation,
            "sep1_hardening": state.rules.sep1_hardening,
            "chance_quality_role": state.rules.chance_quality_role,
            "h2h_role": state.rules.h2h_role,
            "price_can_promote_structure": state.rules.price_can_promote_structure,
        },
        "change_control": state.change_control.model_dump(),
        "competition_scope": state.competition_scope.model_dump(),
    }

File: apps/web/lib/model-control.ts
const apiBaseUrl = process.env.API_BASE_URL ?? "http://localhost:8000";

export type ModelControlState = {
  banner: string;
  model: {
    name: string;
    version: string;
    regime: string;
    timezone: string;
  };
  rules: {
    recent_total_leakage_confirmation: boolean;
    sep1_hardening: boolean;
    chance_quality_role: string;
    h2h_role: string;
    price_can_promote_structure: boolean;
  };
  change_control: {
    audit_can_modify_model: boolean;
    silent_rule_changes: boolean;
    explicit_user_approval_required: boolean;
    production_requires_canonical_state: boolean;
  };
  competition_scope: Record<string, unknown>;
};

export async function getModelControlState(): Promise<ModelControlState> {
  const response = await fetch(`${apiBaseUrl}/api/v1/model/state`, {
    cache: "no-store",
  });
  if (!response.ok) {
    throw new Error(`Model state API returned ${response.status}`);
  }
  return (await response.json()) as ModelControlState;
}
